<center><img src='Images/cao_image.png'></center>

# A Comparison Analysis of CAO points for 2019, 2020, 2021
Source: [CAO Webpage](http://www.cao.ie/index.php?page=mediastats)
***

<br>

## Table of Contents


#### [1. Introduction](#Intro)

#### [2. Retrieving the Data](#)
&nbsp;&nbsp;&nbsp;&nbsp;[- 2021 Points](#2021Points)<br>
&nbsp;&nbsp;&nbsp;&nbsp;[- 2020 Points](#2020Points)<br>
&nbsp;&nbsp;&nbsp;&nbsp;[- 2019 Points](#2019Points)<br>


#### [3. Concatenating & Joining the Data Sets](#)
&nbsp;&nbsp;&nbsp;&nbsp;[- ](#)<br>

#### [4. Analysis](#)
&nbsp;&nbsp;&nbsp;&nbsp;[- ](#)<br>

#### [5. ](#)
&nbsp;&nbsp;&nbsp;&nbsp;[- ](#)<br>

<br>

***
# <center> 1. Introduction</center>
***

<br>

***
# <center> 2. Retrieving the Data</center>
***

<br>

### Import Libraries

***



In [4]:
# Import libraries
import pandas as pd

# Create plots.
import matplotlib.pyplot as plt

# Nice plot style. 
import seaborn as sns

# Numerical operations. 
import numpy as np

# Regular expressions.
import re

# HTTP requests.
import requests as rq

# To get dates and times.
import datetime as dt

# Opening URLs.
import urllib.request as urlrq

# Engine to read in excel file.
import openpyxl as oxl

# Ensures plots are shown.
%matplotlib inline

<br>

### Funtion to retrieve current date and time
***

This following function will create a time-stamp when called will be saved as part of a filename to know when exactly the data was retrieved and stored.

In [5]:
def time():
    # Gets the current date and time
    cur_time = dt.datetime.now()
    # Format as a string
    current_time = cur_time.strftime('%Y%m%d_%H%M%S')
    return current_time

<br>

***

# 2021 CAO Data

[Source](http://www.cao.ie/index.php?page=points&p=2021&bb=points)

***


<br>

When first starting this report, the 2021 points were only available in a HTML format in the form of round 1 and round 2 data. Later an excel file became available with data specifying the Interview/Portfolio and EOS midpoints. Because of this, the first part of this section will demonstrate the extraction of the round 1 and round 2 data from the HTML, while the latter part of this section will demonstrate the retrieval of Interview/Portfolio and EOS Midpoints columns from the excel file. 

<br>

### Retrieve data from webserver
***

In [6]:
# Retrieve CAO points from the webserver.
response = rq.get('http://www2.cao.ie/points/l8.php')

# Response 200 signifies a successful request/response.
response

<Response [200]>

<br>

### Save the original data
***

In [7]:
# Creates a file path for the original data
pathHTML = 'data/cao2021' + time() + '.html'  # Note the importance of the filename and how it
                                              # will be easy to find in folders/sorted appropriately

In [8]:
# Saves the original html file.
with open(pathHTML, 'w') as f:
    f.write(response.text)

<br>

### Charset error on server

***


Technically, server states decoding as:

```
    Content-Type: text/html; charset=iso-8859-1.
``` 

However, one line uses \x96 which isn't defined in iso-8859-1. 

Therefore, we use the similar decoding standard cp1252, which is very similar but includes \x96. 

In [9]:
# Server uses incorrect encoding
orig_encoding = response.encoding

# Corrected encoding to cp1252
response.encoding = 'cp1252'

<br>

### Using regular expressions to extract desired data
***

To do: Explain what the regular expression is doing. COnsider doing step-by-step explanation. 

In [10]:
# Compiles the regular expression for matching lines so it doesn't recompile repeatedly.
re_course = re.compile(r'([A-Z]{2}[0-9]{3})  (.*)')  # 'r' python treats string as raw string and doesnt evaluate back slashes
                                                                    # \ {character} means we want the literal character ie., *
                                                                    # ? means 0 or 1 of 
                                                                    # + means 1 or more of 

<br>

### Iterating through the response line by line
***

There is other information (other than course codes, titles & points) that may be useful for analysis. These are the astrisks and hash symbols that specify extra requirements for entry to courses. See images below. 


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img src='images/terms.png'>

Note the location of the symbols in the original HTML file. 


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img src='images/portfolio_interview.png'>
 

The following function was initially created to use on the data retrieved via HTML. While it will not be used because the portfolio/interview data will be taken from the excel spreadsheet instead, it is useful to demonstrate how an algorithm would function to extract the information. 

In [11]:
# Function to separate the *, # and digits in the points string. 
def points_to_list(string):
    portfolio = ''
    if string[0] == '#':
        portfolio = '#'
    random = ''
    if string[-1] == "*":
        random = '*'
    points = ''
    for i in string:
        # Extracts only the digits from the string. 
        if i.isdigit():
            points += i
    return [points, portfolio, random]

<br>

The following code extracts and saves the information we want (course code, course titles, round1 & round 2 points) to a csv file. 

In [12]:
# Create a path for the CSV file.
path2021 = 'data/cao2021' + time() + '.csv'

# Keeping count of the courses we are processing.
course_count = 0

# Open the CSV file for writing.
with open(path2021, 'w') as f:
    # Loops through and prints data from response line by line.
    for line in response.iter_lines():
        # Decoding turns bytes into code points and those code points can be displayed on the screen
        d_line = line.decode('cp1252')
        # Match the string specified in re_course, returning only the courses from the response
        if re_course.fullmatch(d_line):
            # Adds one to the course count
            course_count += 1
            # Extract course code and strip of any white space.
            course_code = d_line[:5].strip()
            # Extract course title and strip it of any white space.
            course_title = d_line[7:57].strip()
            # Points.
            course_points = re.split(' +' , d_line[60:])
            # The last course created 3 substrings w/ line split above & the last substring was subsequently removed.
            # Removing substrings that are not useful. 
            if len(course_points) != 2:
                course_points = course_points[:2]
            # Rejoin the substrings with commas. Because course points is a list we need to specify both items. 
            line_split = [course_code, course_title, course_points[0], course_points[1]]
            f.write(','.join(line_split) + '\n')

<br>

After writing the data to file, I went into the CSV and created a row with headings for each column. 

In [13]:
# Read in 2021 csv file and add header row.
df_2021 = pd.read_csv("data/cao202120211115_190905.csv", 
                names=["course_code", "course_title", "rnd_1", "rnd_2"])    

# Check headings.
df_2021

course_code                                       course_title rnd_1 rnd_2
0         AL801  Software Design for Virtual Reality and Gaming...   300   NaN
1         AL802  Software Design in Artificial Intelligence for...   313   NaN
2         AL803  Software Design for Mobile Apps and Connected ...   350   NaN
3         AL805  Computer Engineering for Network Infrastructur...   321   NaN
4         AL810  Quantity Surveying                            ...   328   NaN
..          ...                                                ...   ...   ...
944       WD211  Creative Computing                            ...   270   NaN
945       WD212  Recreation and Sport Management               ...   262   NaN
946       WD230  Mechanical and Manufacturing Engineering      ...   230   230
947       WD231  Early Childhood Care and Education            ...   266   NaN
948       WD232  Business Information Systems                  ...   261   261

[949 rows x 4 columns]

<br>

## Get EOS and interview/portfolio data
***
We will use the pandas `read_excel()` function to read in the excel file.  <br>
[read_excel() Documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html)

In [14]:
# Save file path to a variable.
file = 'data/cao_point_2021.xlsx'

In [15]:
# Read in data.
eos_2021 = pd.read_excel(file, skiprows=11, engine='openpyxl') # Use openpyxl to open xlsx spreadsheet.

# Check data.
eos_2021.head()

CATEGORY (ISCED Description)                     Course Title  \
0  Engineering and engineering trades  Music and Instrument Technology   
1                              Health              Pharmacy Technician   
2                              Health                   Dental Nursing   
3     Biological and related sciences                  Applied Science   
4         Business and administration                         Business   

  Course Code R1 Points R1 Random R2 Points  R2 Random EOS Points EOS Random  \
0       AL605       211       NaN        NaN       NaN        211        NaN   
1       AL630       308       NaN        NaN       NaN        308        NaN   
2       AL631       311       NaN        NaN       NaN        311        NaN   
3       AL632       297       NaN        NaN       NaN        297        NaN   
4       AL650       AQA       NaN        AQA       NaN        AQA        NaN   

  EOS Midpoints  Course Level                              HEI Test/Interview  \
0           319             6  Athlone Institute of Technology            NaN   
1           409             6  Athlone Institute of Technology            NaN   
2           400             6  Athlone Institute of Technology            NaN   
3           454             6  Athlone Institute of Technology            NaN   
4           351             6  Athlone Institute of Technology            NaN   

   AVP    v  
0  NaN  NaN  
1  NaN  NaN  
2  NaN  NaN  
3  NaN  NaN  
4  avp  NaN

In [16]:
# View column names.
eos_2021.columns

Index(['CATEGORY (ISCED Description)', 'Course Title', 'Course Code',
       'R1 Points', 'R1 Random', 'R2 Points ', 'R2 Random', 'EOS Points',
       'EOS Random', 'EOS Midpoints', 'Course Level', 'HEI', 'Test/Interview',
       'AVP', 'v'],
      dtype='object')

In [17]:
# Delete irrelevant columns.
eos_2021 = eos_2021[['Course Code', 'EOS Midpoints', 'Test/Interview']]

# View the dataframe.
eos_2021

Course Code EOS Midpoints Test/Interview
0          AL605           319            NaN
1          AL630           409            NaN
2          AL631           400            NaN
3          AL632           454            NaN
4          AL650           351            NaN
...          ...           ...            ...
1446       WD211           392            NaN
1447       WD212           304            NaN
1448       WD230           361            NaN
1449       WD231           366            NaN
1450       WD232           329            NaN

[1451 rows x 3 columns]

<br>

### Concatenate and join tables
***

In [18]:
# Set the index to be the course code as this is what we will join the dataframes on.
eos_2021.set_index('Course Code', inplace=True)
df_2021.set_index('course_code', inplace=True)

In [19]:
# Join dataframes.
df_2021 = df_2021.join(eos_2021)

In [20]:
# Check the dataframe.
df_2021

course_title rnd_1 rnd_2  \
course_code                                                                  
AL801        Software Design for Virtual Reality and Gaming...   300   NaN   
AL802        Software Design in Artificial Intelligence for...   313   NaN   
AL803        Software Design for Mobile Apps and Connected ...   350   NaN   
AL805        Computer Engineering for Network Infrastructur...   321   NaN   
AL810        Quantity Surveying                            ...   328   NaN   
...                                                        ...   ...   ...   
WD211        Creative Computing                            ...   270   NaN   
WD212        Recreation and Sport Management               ...   262   NaN   
WD230        Mechanical and Manufacturing Engineering      ...   230   230   
WD231        Early Childhood Care and Education            ...   266   NaN   
WD232        Business Information Systems                  ...   261   261   

            EOS Midpoints Test/Interview  
course_code                               
AL801                 359            NaN  
AL802                 381            NaN  
AL803                 398            NaN  
AL805                 381            NaN  
AL810                 377            NaN  
...                   ...            ...  
WD211                 392            NaN  
WD212                 304            NaN  
WD230                 361            NaN  
WD231                 366            NaN  
WD232                 329            NaN  

[949 rows x 5 columns]

<br>

### Check data integrity
***

In [21]:
# Count the rows in the dataframe.
len(df_2021)

949

On the 12-11-2021, it was verified that 949 courses were documented in the 2021 CAO data online at the following webpage http://www2.cao.ie/points/l8.php. This corresponds with the number of courses in our dataset ensuring that there was no loss of data.

<br>

***

# 2020 CAO data

[Source](http://www.cao.ie/index.php?page=points&p=2020)
***

In [22]:
# Save the URL in a variable.
url_2020 = 'http://www2.cao.ie/points/CAOPointsCharts2020.xlsx'

<br>

### Save the original file
***

In [23]:
# Create a file path for the original data set.
pathxlsx = 'data/CAO2020' + time() + '.xlsx'

# Opening URL
urlrq.urlretrieve(url_2020, pathxlsx)

('data/CAO202020211214_092227.xlsx', <http.client.HTTPMessage at 0x122593d90>)

<br>

### Load data with pandas
[Pandas documentation for reading in excel data](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html)
***

The function from the pandas library to read in an excel file is `pandas.read_excel()`. 

One of this function's parameters is `engine`. If we set the value to openpyxl, we can use this library to open newer excel file formats.

In [24]:
# Load and parse spread sheet. 
df_2020 = pd.read_excel(pathxlsx, skiprows=10, engine='openpyxl') # Use openpyxl to open xlsx spreadsheet.
df_2020

CATEGORY (i.e.ISCED description)  \
0                           Business and administration   
1                         Humanities (except languages)   
2                                                  Arts   
3                                                  Arts   
4                                                  Arts   
...                                                 ...   
1459                       Manufacturing and processing   
1460  Information and Communication Technologies (ICTs)   
1461  Information and Communication Technologies (ICTs)   
1462                                  Personal services   
1463                 Engineering and engineering trades   

                                           COURSE TITLE COURSE CODE2  \
0                                International Business        AC120   
1                                          Liberal Arts        AC137   
2      First Year Art & Design (Common Entry,portfolio)        AD101   
3     Graphic Design and Moving Image Design (portfo...        AD102   
4     Textile & Surface Design and Jewellery & Objec...        AD103   
...                                                 ...          ...   
1459                          Manufacturing Engineering        WD208   
1460                       Software Systems Development        WD210   
1461                                 Creative Computing        WD211   
1462                    Recreation and Sport Management        WD212   
1463           Mechanical and Manufacturing Engineering        WD230   

     R1 POINTS R1 Random * R2 POINTS R2 Random*       EOS EOS Random *  \
0          209         NaN       NaN        NaN       209          NaN   
1          252         NaN       NaN        NaN       252          NaN   
2     #+matric         NaN       NaN        NaN  #+matric          NaN   
3     #+matric         NaN       NaN        NaN  #+matric          NaN   
4     #+matric         NaN       NaN        NaN  #+matric          NaN   
...        ...         ...       ...        ...       ...          ...   
1459       188         NaN       NaN        NaN       188          NaN   
1460       279         NaN       NaN        NaN       279          NaN   
1461       271         NaN       NaN        NaN       271          NaN   
1462       270         NaN       NaN        NaN       270          NaN   
1463       253         NaN       NaN        NaN       253          NaN   

     EOS Mid-point  ...  avp    v Column1 Column2 Column3  Column4  Column5  \
0              280  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1              270  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
2         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
3         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
4         #+matric  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
...            ...  ...  ...  ...     ...     ...     ...      ...      ...   
1459           339  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1460           337  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1461           318  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1462           349  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   
1463           369  ...  NaN  NaN     NaN     NaN     NaN      NaN      NaN   

      Column6  Column7  Column8  
0         NaN      NaN      NaN  
1         NaN      NaN      NaN  
2         NaN      NaN      NaN  
3         NaN      NaN      NaN  
4         NaN      NaN      NaN  
...       ...      ...      ...  
1459      NaN      NaN      NaN  
1460      NaN      NaN      NaN  
1461      NaN      NaN      NaN  
1462      NaN      NaN      NaN  
1463      NaN      NaN      NaN  

[1464 rows x 23 columns]

<br>

### Delete irrelevant columns for our analysis
***

In [25]:
# Delete irrelevant columns.
df_2020.drop(columns=['COURSE TITLE', 
                    'CATEGORY (i.e.ISCED description)', 
                    'R1 Random *', 
                    'R2 Random*',
                    'LEVEL',
                    'EOS', 
                    'EOS Random *',  
                    'HEI', 
                    'avp', 
                    'v', 
                    'Column1', 'Column2', 'Column3', 'Column4', 'Column5', 'Column6', 
                    'Column7', 'Column8'])


COURSE CODE2 R1 POINTS R2 POINTS EOS Mid-point Test/Interview #
0           AC120       209       NaN           280              NaN
1           AC137       252       NaN           270              NaN
2           AD101  #+matric       NaN      #+matric                #
3           AD102  #+matric       NaN      #+matric                #
4           AD103  #+matric       NaN      #+matric                #
...           ...       ...       ...           ...              ...
1459        WD208       188       NaN           339              NaN
1460        WD210       279       NaN           337              NaN
1461        WD211       271       NaN           318              NaN
1462        WD212       270       NaN           349              NaN
1463        WD230       253       NaN           369              NaN

[1464 rows x 5 columns]

<br>

### Spot check the data
***

In [26]:
# Checking random row to ensure data integrity.
df_2020.iloc[650]

CATEGORY (i.e.ISCED description)                                             Arts
COURSE TITLE                        Arts (Drama, Theatre and Performance Studies)
COURSE CODE2                                                                GY118
R1 POINTS                                                                     451
R1 Random *                                                                   NaN
R2 POINTS                                                                     NaN
R2 Random*                                                                    NaN
EOS                                                                           451
EOS Random *                                                                  NaN
EOS Mid-point                                                                 492
LEVEL                                                                           8
HEI                                        National University of Ireland, Galway
Test/Interview #

In [27]:
# Checking the last row.
df_2020.iloc[-1]


CATEGORY (i.e.ISCED description)          Engineering and engineering trades
COURSE TITLE                        Mechanical and Manufacturing Engineering
COURSE CODE2                                                           WD230
R1 POINTS                                                                253
R1 Random *                                                              NaN
R2 POINTS                                                                NaN
R2 Random*                                                               NaN
EOS                                                                      253
EOS Random *                                                             NaN
EOS Mid-point                                                            369
LEVEL                                                                      8
HEI                                        Waterford Institute of Technology
Test/Interview #                                                         NaN

In [28]:
# Saving the data to a CSV file.
path2020 = 'data/cao2020_' + time() + '.csv'

<br>

The spot check proved that the data has maintained its integrity throughout its processing. 

On 8 December 2021, it was verified that there were 1464 courses listed on the CAO webpage for the year 2020. This corresponds to the number of courses in our data set. 

<br>

***
# 2019 CAO data
 
[Source](http://www2.cao.ie/points/lvl8_19.pdf)

***

<br>

## Convert pdf to csv
***

<b>How was the data was prepared?</b>

- Copied the content from the pdf (in the Preview application)
- Pasted into a Word document so that it formatted nicely
- Then copied the data from the Word document to a csv file
- Deleted preamble and unnecessary data such as page numbers and Institution names while keeping the course code and points etc.
- The saved csv file was then read in with the following code

In [29]:
# Read in 2019 cao csv file & use the tab character as the delimiter.
df_2019 = pd.read_csv('data/cao2019_20211101_213010.csv', sep='\t')
df_2019

Course Code                             INSTITUTION and COURSE   EOS   \
0         AL801    Software Design with Virtual Reality and Gaming   304    
1         AL802               Software Design with Cloud Computing   301    
2         AL803   Software Design with Mobile Apps and Connected...  309    
3         AL805        Network Management and Cloud Infrastructure   329    
4         AL810                                 Quantity Surveying   307    
..           ...                                                ...   ...   
925       WD200                                     Arts (options)   221    
926       WD210                       Software Systems Development   271    
927       WD211                                 Creative Computing   275    
928       WD212                    Recreation and Sport Management   274    
929       WD230           Mechanical and Manufacturing Engineering   273    

     Mid   
0    328   
1    306   
2    337   
3    442   
4    349   
..    ...  
925  296   
926  329   
927  322   
928  311   
929  348   

[930 rows x 4 columns]

<br>

### Dealing with white space
***

White space can cause problems when processing & interacting with data, so below we are using the `str.strip()` method from the Pandas library to remove it. <br>

[Pandas strip method documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.strip.html)

In [30]:
# Check columns to find white space. 
df_2019.columns

Index(['Course Code ', 'INSTITUTION and COURSE ', 'EOS ', 'Mid '], dtype='object')

In [31]:
# Remove white space in column titles.
df_2019.columns = df_2019.columns.str.strip()

As seen above there is white space at the end of the course code and course title strings. It is important to move any white space now as it will impede access to the data later on. 


In [32]:
# Strip data of any white space.
df_2019['Course Code'] = df_2019['Course Code'].str.strip()
df_2019['INSTITUTION and COURSE'] = df_2019['INSTITUTION and COURSE'].str.strip()

<br>

***
# <center> 3. Concatenate & Join Data Sets </center>
***

#### In this section, I will join the data from 2021, 2020 and 2019 to a single dataframe.

<br>

<i>First we must:</i>

- <b>Extract and rename the common column names</b> of the dataframes (course code & course title) and create a new dataframe for each year (this will avoid duplication of columns when the tables are joined). 

<i>Next:</i>
- <b>Concatenate all courses</b> to a dataframe called all_courses using the [Pandas concat() function](https://pandas.pydata.org/docs/reference/api/pandas.concat.html).

<i>Then:</i>
- First, we will <b>check for duplicate rows</b>, then we will remove any duplicates. Duplicates can appear in this data if there is even a minor change in the course title, for instance.

<i>Finally:</i>
- <b>Join the other data</b>, that is applicable to our analysis, to the all_courses dataframe i.e., the points for round 1 & 2, EOS midpoints, portfolio & interview data.

<br>

#### Extract columns & rename
***

In [33]:
# Reset index as I had previously set the index to be course_code when joining the two 2021 tables.
df_2021 = df_2021.reset_index()

In [34]:
# Extract course code & title.
courses2021 = df_2021[['course_code', 'course_title']]

In [35]:
courses2021

course_code                                       course_title
0         AL801  Software Design for Virtual Reality and Gaming...
1         AL802  Software Design in Artificial Intelligence for...
2         AL803  Software Design for Mobile Apps and Connected ...
3         AL805  Computer Engineering for Network Infrastructur...
4         AL810  Quantity Surveying                            ...
..          ...                                                ...
944       WD211  Creative Computing                            ...
945       WD212  Recreation and Sport Management               ...
946       WD230  Mechanical and Manufacturing Engineering      ...
947       WD231  Early Childhood Care and Education            ...
948       WD232  Business Information Systems                  ...

[949 rows x 2 columns]

In [36]:
# Extract course code & title.
courses2020 = df_2020[['COURSE CODE2', 'COURSE TITLE']]

# Change column names.
courses2020.columns = ['course_code', 'course_title']

In [37]:
# Check data.
courses2020

course_code                                       course_title
0          AC120                             International Business
1          AC137                                       Liberal Arts
2          AD101   First Year Art & Design (Common Entry,portfolio)
3          AD102  Graphic Design and Moving Image Design (portfo...
4          AD103  Textile & Surface Design and Jewellery & Objec...
...          ...                                                ...
1459       WD208                          Manufacturing Engineering
1460       WD210                       Software Systems Development
1461       WD211                                 Creative Computing
1462       WD212                    Recreation and Sport Management
1463       WD230           Mechanical and Manufacturing Engineering

[1464 rows x 2 columns]

In [38]:
# View 2019 column names.
df_2019.columns

Index(['Course Code', 'INSTITUTION and COURSE', 'EOS', 'Mid'], dtype='object')

In [39]:
courses2019 = df_2019[['Course Code', 'INSTITUTION and COURSE', ]]
courses2019.columns = ['course_code', 'course_title', ]
courses2019

course_code                                       course_title
0         AL801    Software Design with Virtual Reality and Gaming
1         AL802               Software Design with Cloud Computing
2         AL803  Software Design with Mobile Apps and Connected...
3         AL805        Network Management and Cloud Infrastructure
4         AL810                                 Quantity Surveying
..          ...                                                ...
925       WD200                                     Arts (options)
926       WD210                       Software Systems Development
927       WD211                                 Creative Computing
928       WD212                    Recreation and Sport Management
929       WD230           Mechanical and Manufacturing Engineering

[930 rows x 2 columns]

### Concatenate Data Frames
***

Pandas function to concatenate data frames: 

`pandas.concat()`

[Documentation](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) 


Setting `ignore_index` to True ensures that the indices from the old data frames are not brought into the new data frame. This is to ensure no duplication of indices. For example, we do not want multiple rows all with 0 as their index.

In [40]:
# Concatenating 2021 & 2021 courses. 
all_courses = pd.concat([courses2021, courses2020, courses2019], ignore_index=True)
all_courses

course_code                                       course_title
0          AL801  Software Design for Virtual Reality and Gaming...
1          AL802  Software Design in Artificial Intelligence for...
2          AL803  Software Design for Mobile Apps and Connected ...
3          AL805  Computer Engineering for Network Infrastructur...
4          AL810  Quantity Surveying                            ...
...          ...                                                ...
3338       WD200                                     Arts (options)
3339       WD210                       Software Systems Development
3340       WD211                                 Creative Computing
3341       WD212                    Recreation and Sport Management
3342       WD230           Mechanical and Manufacturing Engineering

[3343 rows x 2 columns]

The number of rows seems larger than it should be, so it's important to check for duplicate rows in the dataframe.

## Manage Duplicate Rows
***

#### To deal with duplicate rows, the following functions from the Pandas library can be used: 

<br>

`pandas.DataFrame.duplicated()`

[Documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html)

We will use the `duplicated()` function to check for duplicate rows based on the course code. As the course code represents a single course, the code should be unique and appear once in the dataset. The `subset` parameter will specify the column on which to check for duplicates.

<br>

`pandas.DataFrame.drop_duplicates()` 

[Documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

The `drop_duplicates()` function along with `subset` set to course_code simply deletes duplicated rows and as default keeps the first record of the duplicates which appears in the dataset. By setting `inplace` to be True it means that the duplicates will be dropped in-place and not returned in a new dataframe.

<br>

In [41]:
# Find extra duplicates based on course code. 
all_courses[all_courses.duplicated(subset=['course_code'])]

course_code                                       course_title
949        AC120                             International Business
950        AC137                                       Liberal Arts
951        AD101   First Year Art & Design (Common Entry,portfolio)
952        AD102  Graphic Design and Moving Image Design (portfo...
953        AD103  Textile & Surface Design and Jewellery & Objec...
...          ...                                                ...
3338       WD200                                     Arts (options)
3339       WD210                       Software Systems Development
3340       WD211                                 Creative Computing
3341       WD212                    Recreation and Sport Management
3342       WD230           Mechanical and Manufacturing Engineering

[1692 rows x 2 columns]

In [42]:
# Remove duplicate rows, implementing changes in-place, while ignoring indices. 
all_courses.drop_duplicates(subset=['course_code'], inplace=True, ignore_index=True)

In [43]:
# Take a look at the data. 
all_courses

course_code                                       course_title
0          AL801  Software Design for Virtual Reality and Gaming...
1          AL802  Software Design in Artificial Intelligence for...
2          AL803  Software Design for Mobile Apps and Connected ...
3          AL805  Computer Engineering for Network Infrastructur...
4          AL810  Quantity Surveying                            ...
...          ...                                                ...
1646       SG441                              Environmental Science
1647       SG446                                Applied Archaeology
1648       TL803                                   Music Technology
1649       TL812                       Computing with Digital Media
1650       WD148                                  Tourism Marketing

[1651 rows x 2 columns]

In [44]:
# Check again for duplicates.
all_courses.duplicated().sum()

0

<b>No duplicates, success! The data set went from 3343 rows to 1651, removing a total of 1692 rows.</b>

<br>

## Join the points & other data
***

<br>

#### In this section:

- we change the column names for each dataframe to reflect their year (so we can identify the data after the tables have been joined) and add these columns to the all_courses dataframe.
- we set the index for the all_courses dataframe to be the course_code - this ensures the data is joined on the (common) course_code and not on the index.
- Finally, when all the tables have been joined together we can do a spot check to ensure data integrity.

<br>

### Join the 2021 data to all_courses dataframe
***

In [45]:
# View column names.
df_2021.columns

Index(['course_code', 'course_title', 'rnd_1', 'rnd_2', 'EOS Midpoints',
       'Test/Interview'],
      dtype='object')

In [46]:
# Change df_2021 column names to include the year.
df_2021.columns = ['course_code', 
                   'course_title', 
                   'points_r1_2021', 
                   'points_r2_2021', 
                   'eos_mid_2021', 
                   'test/interview_2021']

# Check dataframe.
df_2021.head()

course_code                                       course_title  \
0       AL801  Software Design for Virtual Reality and Gaming...   
1       AL802  Software Design in Artificial Intelligence for...   
2       AL803  Software Design for Mobile Apps and Connected ...   
3       AL805  Computer Engineering for Network Infrastructur...   
4       AL810  Quantity Surveying                            ...   

  points_r1_2021 points_r2_2021 eos_mid_2021 test/interview_2021  
0            300            NaN          359                 NaN  
1            313            NaN          381                 NaN  
2            350            NaN          398                 NaN  
3            321            NaN          381                 NaN  
4            328            NaN          377                 NaN

In [47]:
# Join the 2021 points to the all_courses dataframe.
all_courses = all_courses.join(df_2021[['points_r1_2021', 
                                        'points_r2_2021', 
                                        'eos_mid_2021', 
                                        'test/interview_2021']])

<br>

### Setting the index to be course code 
***

As mentioned above, setting the index to course_code ensures the data is joined on the (common) course_code and not on the index.

In [48]:
# Set the all_courses index to be the course_code column.
all_courses = all_courses.set_index('course_code')

In [49]:
# Check.
all_courses.head()

course_title points_r1_2021  \
course_code                                                                     
AL801        Software Design for Virtual Reality and Gaming...            300   
AL802        Software Design in Artificial Intelligence for...            313   
AL803        Software Design for Mobile Apps and Connected ...            350   
AL805        Computer Engineering for Network Infrastructur...            321   
AL810        Quantity Surveying                            ...            328   

            points_r2_2021 eos_mid_2021 test/interview_2021  
course_code                                                  
AL801                  NaN          359                 NaN  
AL802                  NaN          381                 NaN  
AL803                  NaN          398                 NaN  
AL805                  NaN          381                 NaN  
AL810                  NaN          377                 NaN

<br>

### Join the 2020 points to the all_courses dataframe
***

In [50]:
# Have a look at the columns.
df_2020.columns

Index(['CATEGORY (i.e.ISCED description)', 'COURSE TITLE', 'COURSE CODE2',
       'R1 POINTS', 'R1 Random *', 'R2 POINTS', 'R2 Random*', 'EOS',
       'EOS Random *', 'EOS Mid-point', 'LEVEL', 'HEI', 'Test/Interview #',
       'avp', 'v', 'Column1', 'Column2', 'Column3', 'Column4', 'Column5',
       'Column6', 'Column7', 'Column8'],
      dtype='object')

In [51]:
# Extract desired columns. 
df_2020 = df_2020[['COURSE CODE2', 'R1 POINTS', 'R2 POINTS', 'EOS Mid-point', 'Test/Interview #']]

# Change column names.
df_2020.columns = ['course_code', 'points_r1_2020', 'points_r2_2020', 'eos_mid_2020', 'test/interview_2020']

In [52]:
# Set index to be course_code column. 
df_2020.set_index('course_code', inplace=True)

In [53]:
# Join 2020 points to all_courses dataframe.
all_courses = all_courses.join(df_2020)

# Check that the points were added and if column names were changed correctly.
all_courses.head()

course_title points_r1_2021  \
course_code                                                                     
AL801        Software Design for Virtual Reality and Gaming...            300   
AL802        Software Design in Artificial Intelligence for...            313   
AL803        Software Design for Mobile Apps and Connected ...            350   
AL805        Computer Engineering for Network Infrastructur...            321   
AL810        Quantity Surveying                            ...            328   

            points_r2_2021 eos_mid_2021 test/interview_2021 points_r1_2020  \
course_code                                                                  
AL801                  NaN          359                 NaN            303   
AL802                  NaN          381                 NaN            332   
AL803                  NaN          398                 NaN            337   
AL805                  NaN          381                 NaN            333   
AL810                  NaN          377                 NaN            319   

            points_r2_2020 eos_mid_2020 test/interview_2020  
course_code                                                  
AL801                  NaN          367                 NaN  
AL802                  NaN          382                 NaN  
AL803                  NaN          360                 NaN  
AL805                  NaN          360                 NaN  
AL810                  NaN          352                 NaN

<br>

### Join the 2019 points to the all_courses dataframe
***

In [54]:
# View column names.
df_2019.columns

Index(['Course Code', 'INSTITUTION and COURSE', 'EOS', 'Mid'], dtype='object')

In [55]:
# Extract desired columns. 
df_2019 = df_2019[['Course Code', 'Mid', 'EOS']]

# Change column names.
df_2019.columns = ['course_code', 'eos_mid_2019', 'eos_2019']

In [56]:
# Set the index for df_2019 to be the course_code column. 
df_2019.set_index('course_code', inplace=True)

In [57]:
# Join 2019 points to all_courses dataframe.
all_courses = all_courses.join(df_2019)

# Check.
all_courses.head()

course_title points_r1_2021  \
course_code                                                                     
AL801        Software Design for Virtual Reality and Gaming...            300   
AL802        Software Design in Artificial Intelligence for...            313   
AL803        Software Design for Mobile Apps and Connected ...            350   
AL805        Computer Engineering for Network Infrastructur...            321   
AL810        Quantity Surveying                            ...            328   

            points_r2_2021 eos_mid_2021 test/interview_2021 points_r1_2020  \
course_code                                                                  
AL801                  NaN          359                 NaN            303   
AL802                  NaN          381                 NaN            332   
AL803                  NaN          398                 NaN            337   
AL805                  NaN          381                 NaN            333   
AL810                  NaN          377                 NaN            319   

            points_r2_2020 eos_mid_2020 test/interview_2020 eos_mid_2019  \
course_code                                                                
AL801                  NaN          367                 NaN         328    
AL802                  NaN          382                 NaN         306    
AL803                  NaN          360                 NaN         337    
AL805                  NaN          360                 NaN         442    
AL810                  NaN          352                 NaN         349    

            eos_2019  
course_code           
AL801           304   
AL802           301   
AL803           309   
AL805           329   
AL810           307

<br>

### Spot check
***

At this stage it is a good idea to do a spot check to ensure that the joins took place correctly and the data maintained its integrity.

The Pandas has a handy function called `sample()` that randomly selects a single row from the data set.

In [58]:
# Pick random row to check.
all_courses.sample()

# Find a specific course by index.
all_courses.loc[all_courses.index == 'TR263']

course_title points_r1_2021  \
course_code                                                                     
TR263        English Literature/History of Art and Architec...            NaN   

            points_r2_2021 eos_mid_2021 test/interview_2021 points_r1_2020  \
course_code                                                                  
TR263                  NaN          NaN                 NaN            498   

            points_r2_2020 eos_mid_2020 test/interview_2020 eos_mid_2019  \
course_code                                                                
TR263                  497          533                 NaN          NaN   

            eos_2019  
course_code           
TR263            NaN

***
# End